### PR TITLE
[Docs] Note on running metric in dp

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/4353
Metrics needs to be treated differently in DP mode (until we have statefull DP #1895). This PR adds a note to the docs what the user can do for now.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.    
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
Great catch ! Any idea on how to solve ?

**Comment by SkafteNicki** [comment]
@tchaton Only solution is to make a custom version of `DataParallel` (as PR #1895 tried to do) where we can insert hooks between these lines of code:
https://github.com/pytorch/pytorch/blob/73e121de1c69bba6863f38bea4b0044a920fdb20/torch/nn/parallel/data_parallel.py#L167-L168
such that metrics can accumulate their internal state after the data is send through the model, but before the results are gathered.

**Review by SeanNaren — APPROVED** [review]

**Review by tchaton — APPROVED** [review]

File: docs/source/metrics.rst [hunk 1]
```diff
@@ -78,6 +78,26 @@ If ``on_epoch`` is True, the logger automatically logs the end of epoch metric v
         self.valid_acc(logits, y)
         self.log('valid_acc', self.valid_acc, on_step=True, on_epoch=True)
 
+.. note::
+    If using metrics in data parallel mode (dp), the metric update/logging should be done
+    in the ``<mode>_step_end`` method (where ``<mode>`` is either ``training``, ``validation``
+    or ``test``). This is due to metric states else being destroyed after each forward pass,
+    leading to wrong accumulation. In practise do the following:
```
Comment (ydcjeff) [inline docs/source/metrics.rst:85]:
```suggestion
    leading to wrong accumulation. In practice do the following:
```

**Review by ydcjeff — APPROVED** [review]

**Review by justusschock — APPROVED** [review]
